### DISCUSSION TITLE
Mesa running on pycafe (pyodide 'deployment')

### DISCUSSION THREAD

**Discussion by maartenbreddels** [mesa/mesa#2170] [Show and tell]
Hi all,

recently we pre-launched https://py.cafe - a platform to run, edit and share Python apps in the browser.
Also, as the creator of Solara, I was curious if Mesa worked on this platform, and it works nicely:

See: https://py.cafe/maartenbreddels/mesa-example-wealth

(Note: we have the option to store a preview image as well, which is ideal for docs or social sharing - the image URL is https://py.cafe/preview/maartenbreddels/mesa-example-wealth )

So you can create links with a preview image like this:
[![preview](https://py.cafe/preview/maartenbreddels/mesa-example-wealth)](https://py.cafe/maartenbreddels/mesa-example-wealth)

We use it for solara to communicate with users on how to do things, like share examples. We've also been working with other open source libraries to use it as a building block for documentation as the apps can be easily embedded in an iframe.

```
<iframe src="https://py.cafe/embed/maartenbreddels/mesa-example-wealth" width="100%" height="400px" border="0"></iframe>
```

It is also easy to 'extract' the code, to inject it in your docs, e.g.:

https://py.cafe/files/maartenbreddels/mesa-example-wealth/app.py

Let me know if the mesa developers find this useful, and if so, in what ways. We are happy to listen to ideas and feedback to improve py.cafe.

Regards,

Maarten Breddels

PS: it's not an open source product, but we will make sure the free usage is sufficient for open source usage.

**Comment by maartenbreddels** [comment]
FYI: see https://github.com/mckinsey/vizro/pull/569 for how this project will use pycafe

**Comment by EwoutH** [comment]
Looks very interesting! We have been discussing some kind of equivalent to [NetLogo Web](https://www.netlogoweb.org/) for some time, this definitely looks like one of the more promising solutions. Native Solara and Plotly support is definitively a pre.

CC @rht @tpike3 @Corvince

**Reply by rht** [reply]
Looks promising! I will do more extensive testing later. Been swamped at work.
1. I noticed that Solara has to be installed from scratchevery time, if this could be cached (or could be faster via `uv`? not sure how Pyodide `pip install` works).
2. Due to the limited space available on my laptop screen, the user controls are displayed on top of the simulation plots.
3. Can it render folder hosted on GitHub, e.g. [virus on network](https://github.com/projectmesa/mesa-examples/tree/main/examples/virus_on_network)?
![2024-07-09T01:36:38,807687806-04:00](https://github.com/projectmesa/mesa/assets/395821/72c51347-c6fe-4707-96fa-72ad20856744)


@Corvince made an example of sharing a model via Jupyter Lite / Pyodide: https://github.com/projectmesa/mesa/discussions/1954, and comparing the UI, I think it is more hectic than py.cafe's. And I assume py.cafe is more lightweight as well? And indeed, with py.cafe, we can deploy something similar to NetLogo Web sooner.

**Reply by maartenbreddels** [reply]
Hi @rht 

Thank you for trying this out, and your feedback.

1. It should not install every time, only when the requirements change. We use uv for resolving the dependencies (although this is an implementation detail). And, downloaded packages will be cached in the browser cache.
2. If you hit the maximize button on the top right next to the inline browser, the app shows without the editor (you should also see the browser URL change, this is also a URL you can share or use in a link)
3. It can kind of fetch a folder, but it's a bit manual, so I created a pycafe snippet that generates a pycafe snippet (yeah, pretty meta/inception): https://py.cafe/maartenbreddels/generate-url-for-mesa-virus-network (Note that we currently do not show directories in the file explorer, but we will work on that.)

**Reply by rht** [reply]
1. I see.
2. Yes it worked. Before you said this, I thought I'd have to modify Mesa's code to detect whether the env is py.cafe and choose the low-width layout. (note: I didn't notice that Solara now has `solara.ColumnsResponsive`, which didn't exist when we did the 2 layouts in the past. This will simplify the code.).
3. That's cool, will have to keep in mind of that capability. I saw that the snippet content is embedded in the URL.

I also measured the RAM usage via Chromium's memory browser. A tab consumes ~350 MB of RAM, about half of Jupyter Lite. Although when I opened the pycafe snippet generator, one of the tabs is at ~600 MB.

**Comment by maartenbreddels** [comment]
Actually, https://github.com/vaexio/vaex/pull/2331 inspired me to look at vaex on wasm. I got it compiling, but it needs pyarrow (which is in the making).

Looking forward to your feedback and also your ideas on how you'd ideally want to use it.

**Reply by tpike3** [reply]
@maartenbreddels py.cafe is very cool! I am working on a similar project I think has overlap with yours, with the general goal of making tech and simulations more accessible. Would you have time for a videochat? Maybe the week of the 22nd?

**Reply by maartenbreddels** [reply]
Hi Tom,

absolutely, would love to chat. Please contact me at maartenbreddels@gmail.com to find a timeslot.

cheers,

Maarten

**Reply by EwoutH** [reply]
If you find a time, and you would allow me, I might be interested in joining, if I'm available.

**Comment by EwoutH** [comment]
I so wish I could just throw a few weeks of engineering time against this problem. Having interactive, no-code, user-facing examples would so helpful for students and modellers to get into the Mesa ecosystem. Like having all our [examples](https://github.com/projectmesa/mesa-examples) runnable from an online interface without installing anything.

**Reply by maartenbreddels** [reply]
That's a shared dream then :)
We'd like to see projects do this, and we also want to do this for some of our own projects.

**Reply by EwoutH** [reply]
This would actually have been the perfect GSoC project. It's clearly compartmentalised. It's large, but not necessarily very difficult. It has very clear users value.

But no, everybody wants to do AI.

**Comment by rlskoeser** [comment]
Thank you for this discussion! Was looking for a lightweight solution for publishing simulations for a project that is no longer in active development so we can shut down a server. Was able to create a wheel of my project, upload it to PyCafe and add to my requirements, imported all the components of my Solara app from my codebase, and everything seems to work.
 https://py.cafe/app/rlskoeser/simulatingrisk

**Comment by EwoutH** [comment]
Hi @maartenbreddels!

I was working on porting our example visualisations from my experimental [py.cafe demo](https://py.cafe/app/EwoutH/mesa-solara-basic-examples) to our main repo. Is it possible to sync a PyCafe application with (part of) a GitHub repo, so it always uses our latest code and the whole app can be tracked in Git?

**Comment by Sahil-Chhoker** [comment]
Hey @maartenbreddels, we've been running into some issues with Solara in PyCafe lately. It seems that it doesn’t support spinning up multiple threads. Is this an intentional design choice or just a missing feature or not viable? Were you and your team aware of this behavior? Please see the attached image for reference.
<img width="965" height="431" alt="image" src="https://github.com/user-attachments/assets/5a5675d7-9700-4bac-a36f-75cd57fb5bb4" />

Thanks!

**Comment by dietcoke-17** [comment]
@EwoutH i agree with you statement that for every update on the idea on with every launch of new version is it possible to sync the model and run them currently just found out that model are running but visulization arent working is there any 
<img width="1366" height="678" alt="{E77B5FA2-4212-4AD6-84F4-0986AF996977}" src="https://github.com/user-attachments/assets/7b87fdfd-c2e0-4bae-802e-492545a28f31" />
as we can steps are counting but the visulalization isnt working 
<img width="1366" height="681" alt="{DE9FF4A5-3D46-45D1-A304-E30AD524A7BC}" src="https://github.com/user-attachments/assets/ae0a8372-530f-4681-a69a-b73963499d5c" />
here there is some problem with the visualization part just wanted to know architeccture behind it how it renders and let us visualize the models in pycafe

**Comment by dietcoke-17** [comment]
Title: PyCafe/Pyodide compatibility: 7 bugs blocking all 6 basic examples

## Summary
All 6 basic examples in py.cafe are broken. After systematic debugging, here are the causes still not able to find the root causes

## Bugs found

1. Slider class removed (mesa 3.x)
   All 6 app.py files use Slider(...) which no longer exists.
   Fix: replace with SliderInt/SliderFloat dicts.

2. networkx not available in Pyodide by default
   mesa/__init__.py imports discrete_space.network unconditionally,
   which imports networkx at startup — even for non-network models.
   Fix: add networkx to requirements.txt, or make import lazy.

3. matplotlib backend fails in browser
   make_space_component(backend="matplotlib") crashes in Pyodide.
   Fix: backend="altair" works for grid models.

4. mesa.experimental.cell_space removed
   wolf_sheep/agents.py and model.py import from 
   mesa.experimental.cell_space which no longer exists.
   Fix: mesa.discrete_space

5. ABMSimulator requires threading — impossible in Pyodide
   wolf_sheep grass regrowth uses simulator.schedule_event_relative()
   Fix: replace with step-based countdown (implemented and tested)

6. ContinuousSpace not supported by altair backend
   boid_flockers uses ContinuousSpace.
   altair_components.py raises NotImplementedError explicitly.
   Fix needed: matplotlib→PNG→base64 workaround or native support

7. RuntimeError: can't start new thread
   SolaraViz runs model steps in background thread.
   Pyodide blocks all thread creation.
   Fix: use_threads=False (available in mesa 3.5.1)

## Tested on
- Mesa 3.5.1, Python 3.12, Pyodide, PyCafe

But still there are some recurrng problems and output is not generated cant we shift to other platform or make a different route which can be sync changed
